File: backend/scraper/gold_scraper.py
import requests
from bs4 import BeautifulSoup
import time
import json

def scrape_gold_rates():
    """
    Scrapes gold rates from Groww website for Mumbai
    Returns list of text content from specified div class
    """
    # URL of the page
    url = "https://groww.in/gold-rates/gold-rate-today-in-mumbai"
    
    # Headers to mimic browser request
    headers = {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36'
    }
    
    try:
        # Make the request
        response = requests.get(url, headers=headers)
        response.raise_for_status()  # Raise exception for bad status codes
        
        # Parse the HTML
        soup = BeautifulSoup(response.text, 'html.parser')
        
        # Find all divs with the specified class
        gold_rate_divs = soup.find_all('div', class_='grp846CardDiv backgroundPrimary borderPrimary valign-wrapper vspace-between')
        
        # Extract text from each div
        results = []
        for div in gold_rate_divs:
            div_text = div.get_text(strip=True, separator=' ')
            results.append(div_text)
            
        return results
        
    except requests.RequestException as e:
        print(f"Error fetching the webpage: {e}")
        return None
    except Exception as e:
        print(f"Error parsing the webpage: {e}")
        return None

def write_json_file(data):
    f = open("gold_rates.json", "w")
    json.dump(data, f, ensure_ascii=True, indent=4)
    f.close()

def main():
    # Add delay to be respectful to the website
    time.sleep(2)
    # Scrape the rates
    gold_rates = scrape_gold_rates()
    rates_parsed = []
    for object in gold_rates:
        split_object = object.split()
        data = {"purity":split_object[0], 
                "date": " ".join(split_object[4:7]),
                "currency":"INR", # TODO: HARDCODED VALUE
                "rate_per_10gm":split_object[8],
                "rate_per_1gm":round(float(split_object[8])/10,4),
                "fluctuation":split_object[11]
                }
        rates_parsed.append(data)
    write_json_file(rates_parsed)
    
    if gold_rates:
        print("Found Gold Rates:")
        for object in rates_parsed:
            print(object)
    else:
        print("Failed to retrieve gold rates")

if __name__ == "__main__":
    main()

File: lib/modules/home/services/gold_scraper.dart
import 'package:http/http.dart' as http;
import 'package:html/parser.dart' as parser;
import 'dart:convert';
import 'dart:io';
import '../home_exports.dart';

class GoldRateScraper {
  static const String url =
      'https://groww.in/gold-rates/gold-rate-today-in-mumbai';

  Future<List<GoldRate>> scrapeGoldRates() async {
    try {
      final response = await http.get(
        Uri.parse(url),
        headers: {
          'User-Agent':
              'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36'
        },
      );

      if (response.statusCode != 200) {
        throw Exception('Failed to load webpage');
      }

      final document = parser.parse(response.body);
      final goldRateDivs = document.getElementsByClassName(
          'grp846CardDiv backgroundPrimary borderPrimary valign-wrapper vspace-between');

      return goldRateDivs.map((div) {
        final text = div.text.trim().replaceAll(RegExp(r'\s+'), ' ').split(' ');

        // Extract rate (removing ₹ symbol)
        final rateStr = text[5].replaceAll('\'24₹', '');
        final rate = double.parse(rateStr);

        // Extract fluctuation (removing parentheses)
        final fluctuation = text.last;

        return GoldRate(
          purity: text[0],
          date:
              '${DateTime.now().year}', // Using current year since it's not in the text
          currency: 'INR',
          ratePerTenGm: rate,
          ratePerGm: double.parse((rate / 10).toStringAsFixed(4)),
          fluctuation: fluctuation,
        );
      }).toList();
    } catch (e) {
      return [];
    }
  }

  Future<void> writeJsonFile(List<GoldRate> rates) async {
    final file = File('gold_rates.json');
    final jsonData = rates.map((rate) => rate.toJson()).toList();
    await file.writeAsString(json.encode(jsonData));
  }

  Future<void> main() async {
    await Future.delayed(Duration(seconds: 2));
    final goldRates = await scrapeGoldRates();

    if (goldRates.isNotEmpty) {
      await writeJsonFile(goldRates);
    } else {
      // print('Failed to retrieve gold rates');
      // TODO: Add error handling logic here
    }
  }
}
